fix new air quality module

lasers opened the pull request:
Courtesy of @beetleman. https://github.com/ultrabug/py3status/pull/746

This covers all remarks.
Style in compliance with other modules.

* ~~TODO: Replace `oldthing` w/ `py3.request` L130~~
* ~~TODO: Decide on using `{icon}` or `{category}`~~
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;~~and~~
* ~~TODO: Decide on using `icon_*` or `category_*`~~
I decided on `{category}` and `category_*`. Fight me irl.

To make this more well around, we would have to add things that allows users to override things for i8ln... or they could do that in their own copy of  customized `air_quality` module.
* `aqi_policies = ''` -- Override custom policies
* `aqi_categories = ''` -- Override custom categories
* `aqi_colors = ''` -- Override custom colors
* `aqi_url = ''` -- Override URL

Please take a look. Many thanks.

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,190 @@
+        )
+
+    def _color(self, data):
+        key = self._key(data)
```
you are calling this both in `_color()` and `_full_text()` IMHO this makes a) the code harder to follow  b) we are doing more work than we need.  We should just get the key once and pass it to the functions.

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,190 @@
+    color_unhealthy = '#CC0033'
+    color_unhealthy_sensitive = '#FF9933'
+    color_unknown = '#FFFFFF'
+    color_very_unhealthy = '#660099'
```
these `color_...` attributes should die.

can we just keep the defaults somewhere

```
COLOR_DEFAULTS = {
    'hazardous': '#7E0023',
    'moderate': '#FFDE33',
}
```

lasers replied:
I need help with this bit.

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,190 @@
+    (101, 150, 'icon_unhealthy_sensitive'),
+    (151, 200, 'icon_unhealthy'),
+    (201, 300, 'icon_very_unhealthy'),
+    (300, None, 'icon_hazardous')
```
`500 < None` == `False` so this is no good you could use `int('inf')` (infinity),  `500 < int('inf')`  == `True`

values like `(0, 50, 'good')`, might be easier to use.  But this is looking like a threshold to me so really you don't need that second number at all.  It would also simplify `_key()` slightly.

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,190 @@
+    icon_unhealthy = 'Unhealthy'
+    icon_unhealthy_sensitive = 'Unhealthy for Sensitive Groups'
+    icon_unknown = 'Unknown'
+    icon_very_unhealthy = 'Very Unhealthy'
```
`aqi_very_unhealthy`?

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,190 @@
+    {aqi} air quality index
+    {category} health risk category
+    {location} location name/uid
+    {random} output from "random" parameter in `iaqi` array.
```
maybe stick to actual ones though I like the idea of `{random}`

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,170 @@
+    (101, 150, 'unhealthy_sensitive'),
+    (151, 200, 'unhealthy'),
+    (201, 300, 'very_unhealthy'),
+    (301, int('inf'), 'hazardous')
```
needs to be `float('inf')`

lasers commented:
@tobes Please have a look. Commits seems messed up, but if that first original (Beetleman) PR goes in, then I will update the branch.

The `CATEGORY/COLOR/ETC` options are hardcoded right now instead of full `aqi_*` configurations. Do we want to try and make them fully configurable? Or to put it in as-is (fixed).... Might be nice to wait and see if somebody got a different URL / settings that they actually want to use first...  Maybe they make PR w/ that eventually.

tobes commented:
@lasers if you wait a bit I'm actually reviewing it now so hold on a while

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,170 @@
+    rest = path[1:len(path)]
+    if len(path) == 1:
+        return coll.get(first, default)
+    return _get_in(coll.get(first, default), rest, default)
```
This function could move into `Py3status` I don't really like these external functions.

Anyhow it could really do with a docstring explaining what it's doing.

I'm not entirely sure we actually need this function

lasers replied:
I tried that before. Had issues about using `self` and self-reference inside (on line 96). If we don't need this function, then I will need help with converting line 132, 133 to something.

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,170 @@
+            location=self.location,
+            category=category,
+            aqi=aqi,
+            **{k: _v.get('v') for k, _v in iaqi.items()}
```
`_v` -> `v`

lasers replied:
How to convert `**{k: _v.get('v') for k, _v in iaqi.items()}` to work with...
```
        return self.py3.safe_format('location': self.location,
                                    'category': category,
                                    'aqi': aqi,
            **{k: v.get('v') for k, v in iaqi.items()}
        )
```

tobes replied:
something like

        return self.py3.safe_format(self.format,
                               dict(location=self.location,
                                    category=category,
                                    aqi=aqi,
            **{k: v.get('v') for k, v in iaqi.items()})
        )

but maybe do the `{k: v.get('v') for k, v in iaqi.items()}` bit before this function to be more readable

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,170 @@
+        if not color:
+            color = COLOR.get(self._key)
+        if not color:
+            color = None
```
this is not needed.  `COLOR.get(self._key)` will give `None` if not found but you can be more explicit `COLOR.get(self._key, None)` if you wish

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,170 @@
+            category = CATEGORY.get(self._key)
+        if not category:
+            category = STRING_UNKNOWN
+        return self.format.format(
```
`py3.safe_format()`

tobes commented:
We can remove the need for `urlencode` but you need to rebase to the latest master as this needed a bugfix #824

```
diff --git a/py3status/modules/air_quality.py b/py3status/modules/air_quality.py
index 7709b08..f9a62c3 100644
--- a/py3status/modules/air_quality.py
+++ b/py3status/modules/air_quality.py
@@ -51,13 +51,6 @@ air_quality {
 @license BSD
 """
 
-try:
-    # python 3
-    from urllib.parse import urlencode
-except ImportError:
-    # python 2
-    from urllib import urlencode
-
 AQI = (
     (0, 50, 'good'),
     (51, 100, 'moderate'),
@@ -105,13 +98,12 @@ class Py3status:
     def _update_api(self):
         url_path = 'feed/{}'.format(self.location)
         url_token = {'token': self.auth_token}
-        url_full = '{url_base}/{url_path}/?{url_query}'.format(
+        url_full = '{url_base}/{url_path}/'.format(
             url_base=URL,
-            url_path=url_path,
-            url_query=urlencode(url_token)
+            url_path=url_path
         )
         try:
-            self._api_data = self.py3.request(url_full).json()
+            self._api_data = self.py3.request(url_full, params=url_token).json()
         except (self.py3.RequestException):
             self._api_data = None

```

tobes commented on py3status/modules/air_quality.py:
```diff
@@ -0,0 +1,159 @@
+    {aqi} air quality index
+    {category} health risk category
+    {location} location/uid
+    {something} output from "something" parameter in `iaqi` array.
```
I think we should remove `{something}` as it is not a placeholder.

maybe a comment like 
```A station may have their own additional set of custom parameters such as {pm25}, {pm10}, etc. You can find these by using `curl http://api.waqi.info/feed/LOCATION?token=YOUR_TOKEN` and look for them```

tobes commented:
some sample output would be great too

lasers commented:
```
{'color':'#009966', 'full_text':'AQI: 49 Good':}
{'color':'#FFDE33', 'full_text':'AQI: 65 Moderate':}
{'color':'#FF9933', 'full_text':'AQI: 103 Unhealthy for Sensitive Groups':}
{'color':'#CC0033', 'full_text':'AQI: 165 Unhealthy':}
{'color':'#660099', 'full_text':'AQI: 220 Very Unhealthy':}
{'color':'#7E0023', 'full_text':'AQI: 301 Hazardous':}
```

lasers commented:
All done. Some things are still hardcoded, but should be good for now until we get some other air quality URLs to use. If you want to, I can apply patch to remove/update `_get_in` function.

tobes commented:
Because of how this module works with the 'demo' token (we get shanghai data) I think this may confuse some users.  So I think we should show the location of our data.  Maybe we should make the extra data in city available too.  Also document that the 'demo' token does this.

```
diff --git a/py3status/modules/air_quality.py b/py3status/modules/air_quality.py
index d3d40ce..0f33772 100644
--- a/py3status/modules/air_quality.py
+++ b/py3status/modules/air_quality.py
@@ -110,7 +110,7 @@ def _get_in(coll, path, default=None):
 class Py3status:
     auth_token = 'demo'
     cache_timeout = 3600
-    format = 'AQI: {aqi} {category}'
+    format = 'AQI: {location} {aqi} {category}'
     location = 'Shanghai'
 
     def _update_api(self):
@@ -140,6 +140,7 @@ class Py3status:
     def _get_full_text(self):
         aqi = _get_in(self._api_data, ['data', 'aqi'])
         iaqi = _get_in(self._api_data, ['data', 'iaqi'], {})
+        city = _get_in(self._api_data, ['data', 'city'], {})
         try:
             category = getattr(self, 'CATEGORY_{}'.format(self._key).lower())
         except:
@@ -147,7 +148,7 @@ class Py3status:
         if not category:
             category = STRING_UNKNOWN
         return self.py3.safe_format(self.format,
-                                    dict(location=self.location,
+                                    dict(location=city.get('name'),
                                          category=category,
                                          aqi=aqi,
                                          **{k: v.get('v') for k, v in iaqi.items()}))
```

I'd also like that placeholder docstring change :)

But generally I think we are almost done here

lasers commented:
Sorry, I thought I removed it. Must be from copypasta when I was trying to trim the new `Note` and sorted it back in nicely without realizing.

I added `AQI: ` there because it looks nice, but I also added `location` placeholder and kept it out of the format to somewhat preserve original PR. If we want new `location` placeholder, then I'll remove `AQI: ` so it can make more sense and be less annoying.

Is it okay to have name right after SAMPLE OUTPUT? Idk if your code works with that.
```
SAMPLE OUTPUT
aqi_good
{'color':'#009966', 'full_text':'AQI: 49 Good'}
```

Will rename sensitive group. Too long imho.

ultrabug commented:
Thanks a lot everyone ! This is a cool new module 👍
